### PR TITLE
[ALLUXIO-2234] Add DirectS3AOutputStream

### PR DESCRIPTION
**Pull request by calvinjia**
https://alluxio.atlassian.net/browse/ALLUXIO-2234

### CODE REVIEW

**Comment by calvinjia** [comment]
Jenkins, test this please.

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3ADirectOutputStreamTest.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  /**
+   * Tests to ensure the permanent file path is used when upload is called.
+   */
+  @Test
+  public void close() throws Exception {
+    mStream.close();
+    Mockito.verify(mManager).upload(
+        new PutObjectRequest(BUCKET_NAME, KEY, mFile).withMetadata(Mockito
```
Comment (aaudiber) [inline underfs/s3a/src/test/java/alluxio/underfs/s3a/S3ADirectOutputStreamTest.java]: mFile isn't used anywhere but here - how does it get into the PutObjectRequest?

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 2]
```diff
@@ -217,6 +217,17 @@ public static String temporaryFileName(long nonce, String path) {
   }
 
   /**
+   * @param path the path of the file, possibly temporary
+   * @return the permanent path of the file if it was temporary, or the original path if it was not
+   */
+  public static String getPermanentFileName(String path) {
+    if (isTemporaryFileName(path)) {
+      return path.substring(0, path.length() - 31);
```
Comment (madanadit) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java]: i don't like having a magic number 31 here. can we calculate this from the temporary file suffix?

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 3]
```diff
@@ -72,6 +72,7 @@
   UNDERFS_S3_PROXY_PORT(Name.UNDERFS_S3_PROXY_PORT, null),
   UNDERFS_S3_THREADS_MAX(Name.UNDERFS_S3_THREADS_MAX, 22),
   UNDERFS_S3_UPLOAD_THREADS_MAX(Name.UNDERFS_S3_UPLOAD_THREADS_MAX, 2),
+  UNDERFS_S3A_DIRECT_WRITES_ENABLED(Name.UNDERFS_S3A_DIRECT_WRITES_ENABLED, "true"),
```
Comment (gpang) [inline core/common/src/main/java/alluxio/PropertyKey.java]: Does this not work for `s3n`?

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 2 above)
Comment (gpang) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java]: that `31` seems error prone. Can we improve that?

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3ADirectOutputStreamTest.java (hunk 1 above)
Comment (calvinjia) [inline underfs/s3a/src/test/java/alluxio/underfs/s3a/S3ADirectOutputStreamTest.java]: Good catch, this test was not working as intended.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 2 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java]: Yeah I didn't like this either. Do you have any suggestions for calculating the size of the expression? Although our regex is fixed size, most are not so I can see why there is not a size method.

File: core/common/src/main/java/alluxio/PropertyKey.java (hunk 3 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/PropertyKey.java]: This PR only includes the changes for s3a.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 2 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java]: I would like to, do you have any suggestions?

Comment (aaudiber) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java]: something like `String.format("0x%016X", 0).size()`, and we could move `"0x%016X"` to a constant

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3ADirectOutputStreamTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+  /**
+   * Tests to ensure the permanent file path is used when upload is called.
+   */
+  @Test
+  public void close() throws Exception {
+    PutObjectRequest request = Mockito.mock(PutObjectRequest.class);
+
+    Mockito.when(request.withMetadata(Mockito.any(ObjectMetadata.class))).thenReturn(request);
+    PowerMockito.whenNew(PutObjectRequest.class)
```
Comment (aaudiber) [inline underfs/s3a/src/test/java/alluxio/underfs/s3a/S3ADirectOutputStreamTest.java]: Hm, it's not too great to leak implementation details like the constructor we use for `PutObjectRequest` into the unit test.

I'd rather capture the `PutObjectRequest` argument to `mManager.upload()`, and check that it uses the right key. See http://site.mockito.org/mockito/docs/current/org/mockito/Mockito.html#captors

Comment (calvinjia) [inline underfs/s3a/src/test/java/alluxio/underfs/s3a/S3ADirectOutputStreamTest.java]: Updated, thanks.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 2 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java]: I tried to clean this up, but I couldn't find a nice way to link the String format with the regex expression we use to check if its a temp file.

**Comment by calvinjia** [comment]
Jenins, test this please.

**Comment by calvinjia** [comment]
Jenkins, test this please.

**Comment by madanadit** [comment]
LGTM

**Comment by gpang** [comment]
LGTM

**Comment by rxin** [comment]
Wouldn't this lead to data corruption or partial writes?

**Pull request merged**